### PR TITLE
Fixes in Shelley `Imp` framework and other small things

### PR DESCRIPTION
**Pull request by teodanciu**
# Description

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/ImpTest.hs [hunk 1]
```diff
@@ -37,8 +40,22 @@ instance
...
+  ) =>
+  AllegraEraImp era
+
+instance
+  ( Crypto c
+  , NFData (SigDSIGN (DSIGN c))
+  , NFData (VerKeyDSIGN (DSIGN c))
+  , Signable (DSIGN c) (Hash (HASH c) EraIndependentTxBody)
+  ) =>
+  AllegraEraImp (AllegraEra c)
```
Comment (lehins) [inline eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/ImpTest.hs]: What is the purpose of this type class?

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    submitTx_ . mkBasicTx $
+      mkBasicTxBody
+        & inputsTxBodyL @era .~ Set.singleton scriptIn
+        & referenceInputsTxBodyL @era .~ Set.singleton refScriptIn
+
+    -- check that the txin pointing at the reference script output is not spent
+    txinLookup refScriptIn utxo `shouldBe` Just refScriptOut
+
+getUTxO :: ImpTestM era (UTxO era)
+getUTxO = getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
```
Comment (lehins) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs]: I've seen this function before in `eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/LedgerSpec.hs`

We should move it to `eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs` and reuse it in both (and possibly other places)

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 3]
```diff
@@ -576,12 +585,21 @@ addScriptTxWits tx = do
...
+      updatedWits =
+        Map.union -- we exclude the reference scripts, starting with Babbage
+          (toAdd `Map.difference` provided)
+          (wits `Map.difference` provided)
   keyPairs <- mapM impSatisfyNativeScript nativeScripts
-  pure (tx & witsTxL . scriptTxWitsL <>~ scriptsToAdd, mconcat $ catMaybes keyPairs)
+  pure
+    ( tx & witsTxL . scriptTxWitsL .~ updatedWits
+    , mconcat $ catMaybes keyPairs
+    )
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Unnecessary change:
```suggestion
 pure (tx & witsTxL . scriptTxWitsL <>~ scriptsToAdd, mconcat $ catMaybes keyPairs)
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 4]
```diff
@@ -576,12 +585,21 @@ addScriptTxWits tx = do
...
+  let needed = getScriptsNeeded utxo (tx ^. bodyTxL)
+      ScriptsProvided provided = getScriptsProvided utxo tx
+      hashesNeeded = getScriptsHashesNeeded needed
+      toAdd = impScripts `Map.restrictKeys` hashesNeeded
+      nativeScripts = mapMaybe getNativeScript $ Map.elems toAdd
+      wits = tx ^. witsTxL . scriptTxWitsL
+      updatedWits =
+        Map.union -- we exclude the reference scripts, starting with Babbage
+          (toAdd `Map.difference` provided)
+          (wits `Map.difference` provided)
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: We don't need to complicate like that
```suggestion
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 5]
```diff
@@ -576,12 +585,21 @@ addScriptTxWits tx = do
...
   -- TODO: Add proper support for Plutus Scripts
-  let scriptsNeeded = getScriptsNeeded utxo (tx ^. bodyTxL)
-      scriptHashesNeeded = getScriptsHashesNeeded scriptsNeeded
-      scriptsToAdd = impScripts `Map.restrictKeys` scriptHashesNeeded
-      nativeScripts = mapMaybe getNativeScript $ Map.elems scriptsToAdd
+  let needed = getScriptsNeeded utxo (tx ^. bodyTxL)
+      ScriptsProvided provided = getScriptsProvided utxo tx
+      hashesNeeded = getScriptsHashesNeeded needed
+      toAdd = impScripts `Map.restrictKeys` hashesNeeded
+      nativeScripts = mapMaybe getNativeScript $ Map.elems toAdd
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: This is a cleaner approach to achieve the same thing:
```suggestion
      scriptsRequired = impScripts `Map.restrictKeys` hashesNeeded
      scriptsToAdd = scriptsRequired Map.\\ provided
      nativeScripts = mapMaybe getNativeScript $ Map.elems scriptsRequired
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 6]
```diff
@@ -604,42 +622,45 @@ fixupFees ::
   Tx era ->
   Set.Set (KeyHash 'Witness (EraCrypto era)) ->
   ImpTestM era (Tx era)
-fixupFees tx nativeScriptKeyWits = do
+fixupFees _tx nativeScriptKeyWits = do
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Underscores indicate unused, which is clearly not the case
```suggestion
fixupFees tx nativeScriptKeyWits = do
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 7]
```diff
@@ -604,42 +622,45 @@ fixupFees ::
...
   pp <- getsNES $ nesEsL . curPParamsEpochStateL
   utxo <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
   certState <- getsNES $ nesEsL . esLStateL . lsCertStateL
   kpSpending <- lookupKeyPair =<< freshKeyHash
   kpStaking <- lookupKeyPair =<< freshKeyHash
-  (rootTxIn, rootTxOut) <- lookupImpRootTxOut
+  rootTxIn <- fst <$> lookupImpRootTxOut
   let
+    tx = _tx & bodyTxL . inputsTxBodyL %~ Set.insert rootTxIn
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Possibly a name like this instead:
```suggestion
    txWithRoot = tx & bodyTxL . inputsTxBodyL %~ Set.insert rootTxIn
```

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs [hunk 8]
```diff
@@ -604,42 +622,45 @@ fixupFees ::
...
-    outputsTotalCoin = sumAllCoin $ tx ^. bodyTxL . outputsTxBodyL
-    remainingCoin = (rootTxOut ^. coinTxOutL) <-> (outputsTotalCoin <+> deposits <-> refunds)
-    remainingTxOut =
+    withdrawals = fold . unWithdrawals $ tx ^. bodyTxL . withdrawalsTxBodyL
+    consumedTotal = coinBalance (txInsFilter utxo (tx ^. bodyTxL . inputsTxBodyL))
+    outputsTotal = sumAllCoin $ tx ^. bodyTxL . outputsTxBodyL
+    changeBeforeFee =
+      consumedTotal
+        <-> (outputsTotal <+> deposits <-> refunds <-> withdrawals)
+    changeBeforeFeeTxOut =
```
Comment (lehins) [inline eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: We should use "consumed" and "produced" here instead of implementing all the logic all over again. https://github.com/IntersectMBO/cardano-ledger/blob/829386ef553a8d86c29fbd1ede9ad3b3b78b3659/libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs#L231-L247

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs [hunk 9]
```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Test.Cardano.Ledger.Babbage.Imp.TxSpec (
```
Comment (lehins) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs]: Everything in ledger is related to a transaction, so it is not very good organizational approach. Convention that is being established is that test are focused on checking a specific rule. Test in this module seems to be mostly related to `UTXO` rule, so let's rename the module to 
```suggestion
module Test.Cardano.Ledger.Babbage.Imp.UtxoSpec (
```

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+import Cardano.Ledger.Babbage.TxOut (referenceScriptTxOutL)
+
+spec ::
+  forall era.
+  ( AllegraEraImp era
+  , BabbageEraTxBody era
+  ) =>
+  SpecWith (ImpTestState era)
+spec = describe "Tx" $ do
```
Comment (lehins) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs]:
```suggestion
spec = describe "UTXO" $ do
```

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp.hs [hunk 11]
```diff
@@ -30,6 +31,7 @@ spec ::
   Spec
 spec = do
   describe "ConwayImpSpec" $ withImpState @era $ do
+    BabbageTxSpec.spec @era
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp.hs]: Good idea, but not here. For consistency sake, let's add the whole of Babbage imp spec here: https://github.com/IntersectMBO/cardano-ledger/blob/829386ef553a8d86c29fbd1ede9ad3b3b78b3659/eras/conway/impl/test/Main.hs#L30-L32
```suggestion
```

**Review by lehins — COMMENTED** [review]
Looking good. Nice catch on the fee and script witnesses fixups

File: eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/ImpTest.hs (hunk 1 above)
Comment (teodanciu) [reply eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/ImpTest.hs]: I added it initially for this: ` NativeScript era ~ Timelock era` and thought I'd add more things here,  but I'm removing it, since that was not the case, and it's not worth all the noise. I can add this constraint in the signature of the test function itself.

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 5 above)
Comment (teodanciu) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Indeed. Thank you

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 8 above)
Comment (teodanciu) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: So much cleaner with `consumed <-> produced` instead! Thanks

File: eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs (hunk 7 above)
Comment (teodanciu) [reply eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs]: Done

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs (hunk 9 above)
Comment (teodanciu) [reply eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/TxSpec.hs]: I will add the minFee specs in the same module, so I'll specialize it to RefScripts as `UtxoRefScriptSpec`

File: eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxoRefScriptSpec.hs [hunk 12]
```diff
@@ -0,0 +1,80 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Test.Cardano.Ledger.Babbage.Imp.UtxoRefScriptSpec (
```
Comment (lehins) [inline eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxoRefScriptSpec.hs]: No, I really meant that we should have one spec per rule.
```suggestion
module Test.Cardano.Ledger.Babbage.Imp.UtxoSpec (
```

Comment (lehins) [reply eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxoRefScriptSpec.hs]: If the module does grow too big at some point we'll split it up into submodules, until then let's organize all of the Imp tests on the per rule basis

**Comment by teodanciu** [comment]
superseded by https://github.com/IntersectMBO/cardano-ledger/pull/3983

**Comment by teodanciu** [comment]
Reopened to make it easier to reivew

**Review by lehins — APPROVED** [review]
Beautiful
